### PR TITLE
VST2: Fail safely

### PR DESCRIPTION
**Pull request by jsakkine**
If initialization fails, do not try to reinitialize Surge. Free the
synth instance when initialization fails. These were the two root causes
that caused Renoise on Linux to eat all memory and stall the whole DAW.
Now the plugin crashes but Renoise stays alive.

With this commit we can really start fix Linux issues with a nice
edit-compiler-run cycle and do not have to wait for 2-5 minutes for
Renoise to finally crash, and have laptop freezed while waiting.

Signed-off-by: Jarkko Sakkinen <jarkko.sakkinen@iki.fi>

### CODE REVIEW

**Comment by jsakkine** [comment]
**NOTE:** Needs macOS testing at minimum. I can test it tomorrow on MBP.

**Comment by baconpaul** [comment]
Ok I tossed in a couple of comments on a quick read. Let me know if it works mac also

**Comment by esaruoho** [comment]
@baconpaul i tried vst + au with renoise + logic and had no issues.

**Comment by jsakkine** [comment]
@baconpaul, tested too, works OK. Basically this introduces a dead state i.e. VST2 plugin does not try to reinit itself. I wonder, does call to init() need to be scattered to every callback like it is now (almost every)? Do not know VST2 well enough to answer this question, so left calls there. Maybe @kurasu has idea, which is the very first callback coming from a VST2 host. I wonder also, is there a way for VST2 plugin to inform host that the plugin won't start?

**Comment by baconpaul** [comment]
I get that. I was worried about this diff

```
void Vst2PluginInstance::getParameterName(VstInt32 index, char* label)
{
-   if (!initialized)
-     init();
+   init();
+   if (state != INITIALIZED)
+	      return;
```

that changes the behavior of getParameterName (which is called a lot) so that it calls init every time. I think that need an `if( state == UNINITIALIZED )` before the call to `init` to match the current behavior, and that happens in a few spots.

**Comment by baconpaul** [comment]
Also you saw my review comment that state shouldn't be an int but should be a State. That way you can catch errors in switches and things like that.

**Comment by baconpaul** [comment]
And of course you didn't see that. Because I had to press "submit review". I apologize.

File: src/vst2/Vst2PluginInstance.cpp [hunk 1]
```diff
@@ -349,22 +365,28 @@ void Vst2PluginInstance::setParameter(VstInt32 index, float value)
...
+      return 0;
+
    return _instance->getParameter01(_instance->remapExternalApiToInternalId(index));
 }
 
 void Vst2PluginInstance::getParameterName(VstInt32 index, char* label)
 {
-   if (!initialized)
-      init();
+   init();
```
[the remark this answers is not in the document]
Comment (jsakkine) [reply src/vst2/Vst2PluginInstance.cpp]: init() checks whether state is initialized or dead so the call can be unconditional.

File: src/vst2/Vst2PluginInstance.h [hunk 2]
```diff
@@ -85,13 +85,19 @@ class Vst2PluginInstance : public AudioEffectX
...
+   };
+
    SurgeSynthesizer* _instance;
    VstEvent* _eventptr[MAX_EVENTS];
    char _eventbufferdata[EVENTBUFFER_SIZE];
 
    int events_this_block, events_processed;
-   bool initialized;
    void init();
+   unsigned int state;
```
[the remark this answers is not in the document]
Comment (jsakkine) [reply src/vst2/Vst2PluginInstance.h]: C and C++ do does very little to make it a real type (compare to for enums in Rust), except scoping. I think using some integer type (unsigned char, unsigned int) is actually better choice because then you at least know the size.

[the remark this answers is not in the document]
Comment (baconpaul) [reply src/vst2/Vst2PluginInstance.h]: Yeah that’s true at compile and assembly eject time but not true at parse and warn time. If you, say, switch on a typed enum you get warnings about missing conditions. 

Also the state isn’t an int. It is a state. There’s a type theoretic reason for it to be so!

And size sure. But we aren’t streaming this anywhere. And if we did we would cast it to the size we wanted.

[the remark this answers is not in the document]
Comment (jsakkine) [reply src/vst2/Vst2PluginInstance.h]: It is now with enum type.

**Comment by jsakkine** [comment]
Did we BTW have a logger already in place?

**Comment by jsakkine** [comment]
Yeah, based on your questions about init() I realized that, it was a bit half-way. Now with tryInit(), I think things are more clean and easier to understand.

**Comment by baconpaul** [comment]
So I reviewed the code, but now we have fixed automation there's a conflict on this branch. @jsakkine I can resolve build and test this for you if you want? The change with which it is conflicted fixes a super big bug on win and mac vst2 so want to be a tiny bit careful. Let me know if you are happy for me to resolve and merge and push.

**Comment by jsakkine** [comment]
@baconpaul, I'll resolve this now and resolve the remaining issues in #269 later (busy at work atm).

**Comment by jsakkine** [comment]
Rebased.

**Comment by baconpaul** [comment]
OK reviewed the code; also confirmed that it build and I can still load a VST2 on Mac (bitwig) and Windows (Cantible Light) with no problem. So I'm going merge this now.
